File: src/precalculus_polar_curves_medium.py
"""
precalculus - polar_curves (medium)
"""

from problem_utils import *

THETA_ALL = [0, pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, 5*pi/6,
             pi, 7*pi/6, 5*pi/4, 4*pi/3, 3*pi/2, 5*pi/3, 7*pi/4, 11*pi/6]


def generate():
    problem_type = choice([
        'convert_polar_to_rect',
        'convert_rect_to_polar',
        'symmetry_test',
        'area_polar',
        'rose_petals',
        'limacon_properties',
    ])

    if problem_type == 'convert_polar_to_rect':
        equation_choice = choice(['circle_r', 'rcos', 'rsin', 'r_acos', 'r_asin'])

        if equation_choice == 'circle_r':
            a_val = randint(1, 10)
            ans = Eq(x**2 + y**2, a_val**2)
            return problem(
                question=f"Convert the polar equation $r = {a_val}$ to rectangular form.",
                answer=ans,
                difficulty=(1300, 1400),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Use $r^2 = x^2 + y^2$: $r = {a_val} \\Rightarrow r^2 = {a_val**2}$",
                    f"${latex(ans)}$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

        elif equation_choice == 'rcos':
            a_val = nonzero(-8, 8)
            ans = Eq(x, a_val)
            return problem(
                question=f"Convert $r\\cos\\theta = {a_val}$ to rectangular form.",
                answer=ans,
                difficulty=(1300, 1400),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Since $x = r\\cos\\theta$: $x = {a_val}$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

        elif equation_choice == 'rsin':
            b_val = nonzero(-8, 8)
            ans = Eq(y, b_val)
            return problem(
                question=f"Convert $r\\sin\\theta = {b_val}$ to rectangular form.",
                answer=ans,
                difficulty=(1300, 1400),
                topic="precalculus/polar_curves",
                solution=steps(f"Since $y = r\\sin\\theta$: $y = {b_val}$"),
                grading_mode="equivalent",
                answer_type="equation"
            )

        elif equation_choice == 'r_acos':
            a_val = 2 * randint(1, 6)
            ans = Eq(x**2 + y**2, a_val * x)
            return problem(
                question=f"Convert $r = {a_val}\\cos\\theta$ to rectangular form.",
                answer=ans,
                difficulty=(1400, 1550),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Multiply both sides by $r$: $r^2 = {a_val}r\\cos\\theta$",
                    f"Substitute $r^2 = x^2+y^2$, $r\\cos\\theta = x$: ${latex(ans)}$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

        else:  # r_asin
            a_val = 2 * randint(1, 6)
            ans = Eq(x**2 + y**2, a_val * y)
            return problem(
                question=f"Convert $r = {a_val}\\sin\\theta$ to rectangular form.",
                answer=ans,
                difficulty=(1400, 1550),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Multiply both sides by $r$: $r^2 = {a_val}r\\sin\\theta$",
                    f"Substitute $r^2 = x^2+y^2$, $r\\sin\\theta = y$: ${latex(ans)}$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

    elif problem_type == 'convert_rect_to_polar':
        equation_choice = choice(['circle_origin', 'vertical_line', 'horizontal_line', 'circle_x', 'circle_y'])

        if equation_choice == 'circle_origin':
            a_val = randint(1, 9)
            ans = Eq(Symbol('r'), a_val)
            return problem(
                question=f"Convert $x^2 + y^2 = {a_val**2}$ to polar form.",
                answer=ans,
                difficulty=(1300, 1400),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"$x^2+y^2 = r^2 \\Rightarrow r^2 = {a_val**2} \\Rightarrow r = {a_val}$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

        elif equation_choice == 'vertical_line':
            a_val = nonzero(-6, 6)
            theta = Symbol('theta')
            ans = Eq(Symbol('r') * cos(theta), a_val)
            return problem(
                question=f"Convert $x = {a_val}$ to polar form.",
                answer=ans,
                difficulty=(1300, 1400),
                topic="precalculus/polar_curves",
                solution=steps(f"$x = r\\cos\\theta$, so $r\\cos\\theta = {a_val}$"),
                grading_mode="equivalent",
                answer_type="equation"
            )

        elif equation_choice == 'horizontal_line':
            b_val = nonzero(-6, 6)
            theta = Symbol('theta')
            ans = Eq(Symbol('r') * sin(theta), b_val)
            return problem(
                question=f"Convert $y = {b_val}$ to polar form.",
                answer=ans,
                difficulty=(1300, 1400),
                topic="precalculus/polar_curves",
                solution=steps(f"$y = r\\sin\\theta$, so $r\\sin\\theta = {b_val}$"),
                grading_mode="equivalent",
                answer_type="equation"
            )

        elif equation_choice == 'circle_x':
            a_val = 2 * randint(1, 5)
            theta = Symbol('theta')
            ans = Eq(Symbol('r'), a_val * cos(theta))
            return problem(
                question=f"Convert $x^2 + y^2 = {a_val}x$ to polar form.",
                answer=ans,
                difficulty=(1400, 1500),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"$r^2 = {a_val} \\cdot r\\cos\\theta$, divide by $r$: $r = {a_val}\\cos\\theta$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

        else:  # circle_y
            a_val = 2 * randint(1, 5)
            theta = Symbol('theta')
            ans = Eq(Symbol('r'), a_val * sin(theta))
            return problem(
                question=f"Convert $x^2 + y^2 = {a_val}y$ to polar form.",
                answer=ans,
                difficulty=(1400, 1500),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"$r^2 = {a_val} \\cdot r\\sin\\theta$, divide by $r$: $r = {a_val}\\sin\\theta$"
                ),
                grading_mode="equivalent",
                answer_type="equation"
            )

    elif problem_type == 'symmetry_test':
        curve_type = choice(['cos_rose', 'sin_rose', 'cardioid_cos', 'cardioid_sin', 'lemniscate'])
        n = choice([2, 3, 4, 5, 6])
        a_val = randint(1, 5)

        if curve_type == 'cos_rose':
            return problem(
                question=f"Is the polar curve $r = {a_val}\\cos({n}\\theta)$ symmetric about the polar axis (the $x$-axis)?",
                answer=True,
                difficulty=(1400, 1550),
                topic="precalculus/polar_curves",
                solution=steps(
                    "Test: replace $\\theta$ with $-\\theta$",
                    f"$r = {a_val}\\cos({n}(-\\theta)) = {a_val}\\cos({n}\\theta)$: equation unchanged",
                    "Symmetric about polar axis."
                ),
                answer_type="boolean"
            )
        elif curve_type == 'sin_rose':
            return problem(
                question=f"Is the polar curve $r = {a_val}\\sin({n}\\theta)$ symmetric about the line $\\theta = \\pi/2$ (the $y$-axis)?",
                answer=True,
                difficulty=(1400, 1550),
                topic="precalculus/polar_curves",
                solution=steps(
                    "Test: replace $\\theta$ with $\\pi - \\theta$",
                    f"$\\sin(n(\\pi-\\theta)) = \\sin(n\\pi - n\\theta)$: for odd $n$ equals $\\sin(n\\theta)$",
                    "Symmetric about $\\theta = \\pi/2$."
                ),
                answer_type="boolean"
            )
        elif curve_type == 'cardioid_cos':
            return problem(
                question=f"Is the cardioid $r = {a_val}(1 + \\cos\\theta)$ symmetric about the polar axis?",
                answer=True,
                difficulty=(1400, 1500),
                topic="precalculus/polar_curves",
                solution=steps(
                    "Replace $\\theta$ with $-\\theta$: $\\cos(-\\theta) = \\cos\\theta$, equation unchanged",
                    "Symmetric about the polar axis."
                ),
                answer_type="boolean"
            )
        elif curve_type == 'cardioid_sin':
            return problem(
                question=f"Is the cardioid $r = {a_val}(1 + \\sin\\theta)$ symmetric about the line $\\theta = \\pi/2$?",
                answer=True,
                difficulty=(1400, 1500),
                topic="precalculus/polar_curves",
                solution=steps(
                    "Replace $\\theta$ with $\\pi - \\theta$: $\\sin(\\pi-\\theta) = \\sin\\theta$, equation unchanged",
                    "Symmetric about $\\theta = \\pi/2$."
                ),
                answer_type="boolean"
            )
        else:  # lemniscate
            return problem(
                question=f"Is the lemniscate $r^2 = {a_val**2}\\cos(2\\theta)$ symmetric about both axes?",
                answer=True,
                difficulty=(1500, 1600),
                topic="precalculus/polar_curves",
                solution=steps(
                    "Replace $\\theta \\to -\\theta$: $\\cos(-2\\theta) = \\cos(2\\theta)$ ✓ (polar axis)",
                    "Replace $(r,\\theta) \\to (-r,\\theta)$: $(-r)^2 = r^2$ ✓ (pole)",
                    "Symmetric about both axes and the pole."
                ),
                answer_type="boolean"
            )

    elif problem_type == 'area_polar':
        a_val = randint(1, 7)
        curve_choice = choice(['circle', 'semicircle', 'rose2'])

        if curve_choice == 'circle':
            ans = pi * a_val**2
            return problem(
                question=f"Find the area enclosed by the polar curve $r = {a_val}$.",
                answer=ans,
                difficulty=(1400, 1550),
                topic="precalculus/polar_curves",
                solution=steps(
                    "$A = \\dfrac{1}{2}\\int_0^{2\\pi} r^2\\, d\\theta$",
                    f"$= \\dfrac{1}{2} \\cdot {a_val**2} \\cdot 2\\pi = {a_val**2}\\pi$"
                ),
                calculator="scientific"
            )
        elif curve_choice == 'semicircle':
            # Area of r = a from 0 to pi: pi*a^2/2
            ans = pi * a_val**2 / 2
            return problem(
                question=f"Find the area swept by $r = {a_val}$ from $\\theta = 0$ to $\\theta = \\pi$.",
                answer=ans,
                difficulty=(1400, 1550),
                topic="precalculus/polar_curves",
                solution=steps(
                    "$A = \\dfrac{1}{2}\\int_0^{\\pi} {a_val**2}\\, d\\theta = \\dfrac{{{a_val**2}\\pi}}{{2}}$"
                ),
                calculator="scientific"
            )
        else:
            # Area of one petal of r = a*cos(2*theta): pi*a^2/8 per petal
            ans = pi * a_val**2 / 8
            return problem(
                question=f"Find the area enclosed by one petal of the rose $r = {a_val}\\cos(2\\theta)$.",
                answer=ans,
                difficulty=(1500, 1600),
                topic="precalculus/polar_curves",
                solution=steps(
                    "One petal lies in $-\\pi/4 \\leq \\theta \\leq \\pi/4$",
                    f"$A = \\dfrac{{1}}{{2}}\\int_{{-\\pi/4}}^{{\\pi/4}} {a_val**2}\\cos^2(2\\theta)\\, d\\theta$",
                    f"$= \\dfrac{{{a_val**2}\\pi}}{{8}}$"
                ),
                calculator="scientific"
            )

    elif problem_type == 'rose_petals':
        # Identify number of petals
        n = choice([2, 3, 4, 5, 6])
        a_val = randint(1, 6)
        trig_str = choice(['\\cos', '\\sin'])
        petals = n if n % 2 == 1 else 2 * n

        return problem(
            question=f"How many petals does the rose curve $r = {a_val}{trig_str}({n}\\theta)$ have?",
            answer=petals,
            difficulty=(1300, 1500),
            topic="precalculus/polar_curves",
            solution=steps(
                f"For $r = a\\,f(n\\theta)$: if $n$ is odd there are $n$ petals; if $n$ is even there are $2n$ petals",
                f"$n = {n}$ ({'odd' if n % 2 == 1 else 'even'}): $\\mathbf{{{petals}}}$ petals"
            )
        )

    else:
        # Limaçon: identify inner loop vs. cardioid vs. dimpled vs. convex
        a_val = randint(1, 6)
        b_val = randint(1, 6)
        trig_str = choice(['\\cos', '\\sin'])
        ratio = Rational(a_val, b_val)

        if a_val < b_val:
            shape = "limaçon with inner loop"
        elif a_val == b_val:
            shape = "cardioid"
        elif a_val < 2 * b_val:
            shape = "dimpled limaçon"
        else:
            shape = "convex limaçon"

        return problem(
            question=f"Classify the polar curve $r = {a_val} + {b_val}{trig_str}(\\theta)$.",
            answer=shape,
            difficulty=(1450, 1600),
            topic="precalculus/polar_curves",
            solution=steps(
                f"Form $r = a + b\\,f(\\theta)$ with $a = {a_val}$, $b = {b_val}$",
                f"Ratio $a/b = {latex(ratio)}$",
                f"$a < b$: inner loop; $a = b$: cardioid; $b < a < 2b$: dimpled; $a \\geq 2b$: convex",
                f"Classification: **{shape}**"
            )
        )

emit(generate())

File: src/precalculus_polar_curves_medium.jl
# precalculus - polar_curves (medium)
# Generated: 2026-03-08T20:34:17.590350

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script r θ x y begin
    set_topic!("precalculus/polar_curves")
    
    problem_type = choice([
        :convert_polar_to_cartesian,
        :convert_cartesian_to_polar,
        :find_symmetry,
        :find_intersection_points,
        :tangent_line_polar,
        :area_polar_region,
    ])
    
    if problem_type == :convert_polar_to_cartesian
        # Convert polar equation to Cartesian
        curve_type = choice([:circle, :line, :rose])
        
        if curve_type == :circle
            a = nonzero(-8, 8)
            # r = a cos(θ) or r = a sin(θ)
            use_cos = choice([true, false])
            if use_cos
                polar_eq = "r = $(a)\\cos(\\theta)"
                # x² + y² = ax → (x - a/2)² + y² = (a/2)²
                cartesian = expand((x - a//2)^2 + y^2)
                cartesian_rhs = (a//2)^2
                cartesian_str = "\\left(x - \\frac{$(a)}{2}\\right)^2 + y^2 = \\frac{$(a^2)}{4}"
            else
                polar_eq = "r = $(a)\\sin(\\theta)"
                # x² + y² = ay → x² + (y - a/2)² = (a/2)²
                cartesian = expand(x^2 + (y - a//2)^2)
                cartesian_rhs = (a//2)^2
                cartesian_str = "x^2 + \\left(y - \\frac{$(a)}{2}\\right)^2 = \\frac{$(a^2)}{4}"
            end
            
            problem(
                question="Convert the polar equation \\\$$(polar_eq)\\\$ to Cartesian (rectangular) form.",
                answer=cartesian_str,
                difficulty=(1200, 1400),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "Multiply both sides by r: \\\$r^2 = $(a)r$(use_cos ? "\\cos(\\theta)" : "\\sin(\\theta)")\\\$",
                    "Use \\\$r^2 = x^2 + y^2\\\$ and \\\$$(use_cos ? "x = r\\cos(\\theta)" : "y = r\\sin(\\theta)")\\\$",
                    "Result: \\\$x^2 + y^2 = $(a)$(use_cos ? "x" : "y")\\\$",
                    "Complete the square: \\\$$(cartesian_str)\\\$"
                ),
                time=120,
            )
        elseif curve_type == :line
            m = nonzero(-5, 5)
            # θ = arctan(m) or r cos(θ - α) = d
            α_deg = randint(1, 8) * 15
            d = nonzero(-10, 10)
            polar_eq = "r = \\frac{$(d)}{\\cos(\\theta - $(α_deg)^\\circ)}"
            cartesian_str = "Line at distance $(abs(d)) from origin"
            
            problem(
                question="Convert the polar equation \\\$\\theta = \\arctan($(m))\\\$ to Cartesian form.",
                answer="y = $(m)x",
                difficulty=(1100, 1300),
                solution=steps(
                    "Given: \\\$\\theta = \\arctan($(m))\\\$",
                    "Take tangent of both sides: \\\$\\tan(\\theta) = $(m)\\\$",
                    "Use \\\$\\tan(\\theta) = \\frac{y}{x}\\\$",
                    "Result: \\\$\\frac{y}{x} = $(m)\\\$, so \\\$y = $(m)x\\\$"
                ),
                time=90,
            )
        else  # rose
            n = randint(2, 5)
            a = nonzero(-6, 6)
            use_cos = choice([true, false])
            polar_eq = "r = $(a)$(use_cos ? "\\cos" : "\\sin")($(n)\\theta)"
            
            problem(
                question="Identify the number of petals in the rose curve \\\$$(polar_eq)\\\$.",
                answer=n % 2 == 1 ? n : 2*n,
                difficulty=(1000, 1200),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "This is a rose curve with \\\$n = $(n)\\\$",
                    n % 2 == 1 ? "Since n is odd, the number of petals is n = $(n)" : "Since n is even, the number of petals is 2n = $(2*n)",
                    "Answer: $(n % 2 == 1 ? n : 2*n) petals"
                ),
                time=60,
            )
        end
        
    elseif problem_type == :convert_cartesian_to_polar
        # Convert Cartesian to polar
        conversion_type = choice([:line, :circle, :parabola])
        
        if conversion_type == :line
            m = nonzero(-4, 4)
            b = nonzero(-8, 8)
            # y = mx + b
            cartesian_eq = "y = $(m)x + $(b)"
            # r sin(θ) = m·r cos(θ) + b
            # r(sin(θ) - m cos(θ)) = b
            polar_ans = "r = \\frac{$(b)}{\\sin(\\theta) - $(m)\\cos(\\theta)}"
            
            problem(
                question="Convert the Cartesian equation \\\$$(cartesian_eq)\\\$ to polar form.",
                answer=polar_ans,
                difficulty=(1300, 1500),
                solution=steps(
                    "Given: \\\$y = $(m)x + $(b)\\\$",
                    "Substitute \\\$x = r\\cos(\\theta)\\\$ and \\\$y = r\\sin(\\theta)\\\$",
                    "\\\$r\\sin(\\theta) = $(m)r\\cos(\\theta) + $(b)\\\$",
                    "Factor: \\\$r(\\sin(\\theta) - $(m)\\cos(\\theta)) = $(b)\\\$",
                    "Result: \\\$$(polar_ans)\\\$"
                ),
                time=120,
            )
        elseif conversion_type == :circle
            h = randint(-6, 6)
            k = randint(-6, 6)
            radius = randint(2, 7)
            r_sq = radius^2
            # (x-h)² + (y-k)² = r²
            cartesian_eq = "\\left(x - $(h)\\right)^2 + \\left(y - $(k)\\right)^2 = $(r_sq)"
            
            problem(
                question="Convert \\\$$(cartesian_eq)\\\$ to polar form and simplify.",
                answer="r^2 - $(2*h)r\\cos(\\theta) - $(2*k)r\\sin(\\theta) = $(r_sq - h^2 - k^2)",
                difficulty=(1400, 1600),
                solution=steps(
                    "Given: \$(x - $(h))^2 + (y - $(k))^2 = $(r_sq)\\\$",
                    "Expand: \\\$x^2 - $(2*h)x + $(h^2) + y^2 - $(2*k)y + $(k^2) = $(r_sq)\\\$",
                    "Use \\\$x^2 + y^2 = r^2\\\$, \\\$x = r\\cos(\\theta)\\\$, \\\$y = r\\sin(\\theta)\\\$",
                    "\\\$r^2 - $(2*h)r\\cos(\\theta) - $(2*k)r\\sin(\\theta) + $(h^2 + k^2) = $(r_sq)\\\$",
                    "Simplify: \\\$r^2 - $(2*h)r\\cos(\\theta) - $(2*k)r\\sin(\\theta) = $(r_sq - h^2 - k^2)\\\$"
                ),
                time=180,
            )
        else  # parabola
            p = nonzero(-5, 5)
            # x² = 4py
            cartesian_eq = "x^2 = $(4*p)y"
            polar_ans = "r = \\frac{$(4*p)\\sin(\\theta)}{\\cos^2(\\theta)}"
            
            problem(
                question="Convert the parabola \\\$$(cartesian_eq)\\\$ to polar form.",
                answer=polar_ans,
                difficulty=(1500, 1700),
                solution=steps(
                    "Given: \\\$x^2 = $(4*p)y\\\$",
                    "Substitute \\\$x = r\\cos(\\theta)\\\$ and \\\$y = r\\sin(\\theta)\\\$",
                    "\\\$r^2\\cos^2(\\theta) = $(4*p)r\\sin(\\theta)\\\$",
                    "Divide by r (assuming \\\$r \\neq 0\\\$): \\\$r\\cos^2(\\theta) = $(4*p)\\sin(\\theta)\\\$",
                    "Result: \\\$$(polar_ans)\\\$"
                ),
                time=150,
            )
        end
        
    elseif problem_type == :find_symmetry
        # Test symmetry of polar curves
        a = nonzero(-6, 6)
        n = randint(2, 6)
        
        symmetry_type = choice([:x_axis, :y_axis, :origin, :rose])
        
        if symmetry_type == :x_axis
            # r = a cos(nθ) has x-axis symmetry when n is odd
            polar_eq = "r = $(a)\\cos(\\theta)"
            symmetries = ["polar axis (x-axis)"]
            
            problem(
                question="Determine all symmetries of the polar curve \\\$$(polar_eq)\\\$. Choose from: polar axis, line θ=π/2, or origin.",
                answer="polar axis",
                difficulty=(1200, 1400),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "Test polar axis: Replace θ with -θ: \\\$r = $(a)\\cos(-\\theta) = $(a)\\cos(\\theta)\\\$ ✓",
                    "Test θ=π/2: Replace θ with π-θ: \\\$r = $(a)\\cos(\\pi - \\theta) = -$(a)\\cos(\\theta)\\\$ ✗",
                    "Test origin: Replace r with -r: Not equivalent ✗",
                    "Answer: Symmetric about the polar axis"
                ),
                time=120,
            )
        elseif symmetry_type == :y_axis
            polar_eq = "r = $(a)\\sin(\\theta)"
            
            problem(
                question="Determine all symmetries of the polar curve \\\$$(polar_eq)\\\$.",
                answer="θ = π/2",
                difficulty=(1200, 1400),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "Test θ=π/2 line: Replace θ with π-θ: \\\$r = $(a)\\sin(\\pi - \\theta) = $(a)\\sin(\\theta)\\\$ ✓",
                    "Test polar axis: Replace θ with -θ: \\\$r = $(a)\\sin(-\\theta) = -$(a)\\sin(\\theta)\\\$ ✗",
                    "Test origin: Not equivalent ✗",
                    "Answer: Symmetric about the line θ = π/2"
                ),
                time=120,
            )
        elseif symmetry_type == :origin
            polar_eq = "r^2 = $(a^2)\\cos(2\\theta)"
            
            problem(
                question="Show that \\\$$(polar_eq)\\\$ has symmetry about the origin.",
                answer="Replace (r,θ) with (-r,θ+π) to verify origin symmetry",
                difficulty=(1400, 1600),
                solution=steps(
                    "Given: \\\$r^2 = $(a^2)\\cos(2\\theta)\\\$",
                    "For origin symmetry, replace (r, θ) with (-r, θ + π)",
                    "\$(-r)^2 = $(a^2)\\cos(2(\\theta + \\pi))\\\$",
                    "\\\$r^2 = $(a^2)\\cos(2\\theta + 2\\pi) = $(a^2)\\cos(2\\theta)\\\$ ✓",
                    "The curve has origin symmetry"
                ),
                time=150,
            )
        else  # rose
            n_petals = choice([3, 5])
            polar_eq = "r = $(a)\\sin($(n_petals)\\theta)"
            
            problem(
                question="How many lines of symmetry does the rose curve \\\$$(polar_eq)\\\$ have?",
                answer=n_petals,
                difficulty=(1300, 1500),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "This is a rose with n = $(n_petals) (odd)",
                    "An n-petal rose (n odd) has n lines of symmetry",
                    "Each line passes through a petal and the origin",
                    "Answer: $(n_petals) lines of symmetry"
                ),
                time=90,
            )
        end
        
    elseif problem_type == :find_intersection_points
        # Find intersection of two polar curves
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        
        # r = a and r = b cos(θ)
        polar1 = "r = $(abs(a))"
        polar2 = "r = $(2*abs(a))\\cos(\\theta)"
        
        # Intersection: |a| = 2|a| cos(θ) → cos(θ) = 1/2 → θ = π/3, 5π/3
        ans_points = "\\left($(abs(a)), \\frac{\\pi}{3}\\right) \\text{ and } \\left($(abs(a)), \\frac{5\\pi}{3}\\right)"
        
        problem(
            question="Find all intersection points of \\\$$(polar1)\\\$ and \\\$$(polar2)\\\$ for \\\$0 \\leq \\theta < 2\\pi\\\$.",
            answer=ans_points,
            difficulty=(1500, 1700),
            solution=steps(
                "Set the equations equal: \\\$$(abs(a)) = $(2*abs(a))\\cos(\\theta)\\\$",
                "Solve for θ: \\\$\\cos(\\theta) = \\frac{1}{2}\\\$",
                "Solutions in [0, 2π): \\\$\\theta = \\frac{\\pi}{3}, \\frac{5\\pi}{3}\\\$",
                "Check for origin (pole): Both curves pass through origin at different θ values",
                "Intersection points: \$($(abs(a)), \\pi/3)\\\$ and \$($(abs(a)), 5\\pi/3)\\\$, plus the pole"
            ),
            time=180,
        )
        
    elseif problem_type == :tangent_line_polar
        # Find slope of tangent line at a point
        a = nonzero(-6, 6)
        θ_val = choice([0, 45, 60, 90, 120, 135, 180])
        θ_rad = deg2rad(θ_val)

        r_val = a * cos(θ_rad)
        slope_num = -a * sin(θ_rad) * sin(θ_rad) + r_val * cos(θ_rad)
        slope_den = -a * sin(θ_rad) * cos(θ_rad) - r_val * sin(θ_rad)

        if abs(slope_den) < 1e-10
            ans_str = "undefined"
        else
            ans_val = round(slope_num / slope_den, digits=4)
            ans_str = string(ans_val)
        end

        problem(
            question="Find the slope of the tangent line to r = $(a)cos(theta) at theta = $(θ_val) degrees.",
            answer=ans_str,
            difficulty=(1600, 1800),
            solution=steps(
                "Use dy/dx formula for polar curves",
                "dr/dtheta = -$(a)*sin(theta)",
                sol("Answer", ans_str)
            ),
            time=180,
        )
    end
end
